### PR TITLE
Change GADT constrainer to consider all parents of the scrutinee type when upcasting it

### PR DESCRIPTION
**Pull request by Linyxus**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala [hunk 1]
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
...
+      }
+
+      scrut match {
         case scrut: TypeRef if scrut.symbol.isClass =>
-          // we do not infer constraints following from all parents for performance reasons
-          // in principle however, if `A extends B, C`, then `A` can be treated as `B & C`
-          scrut.firstParent
+          // consider all parents
+          val parents = scrut.parents
+          val andType = trace(i"andType of scrut", gadts) {
```
Comment (abgruszecki) [inline compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:113]: This trace doesn't seem overall very useful, could you remove it?

File: compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala [hunk 2]
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
...
+            buildAndType(parents)
+          }
+          constrainPatternType(pat, andType)
```
Comment (abgruszecki) [inline compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:116]: Can you check if `andType` exists, like in `case _` below?

File: compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala [hunk 3]
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
...
+              else allParentsSharedWithPat(tp, sym)
             }
           }
-          val sym = firstParentSharedWithPat(tycon, tycon.symbol.asClass)
-          if (sym.exists) scrut.baseType(sym) else NoType
-        case scrut: TypeProxy => scrut.superType
-        case _ => NoType
+          val allSyms = allParentsSharedWithPat(tycon, tycon.symbol.asClass)
+          val baseClasses = allSyms map scrut.baseType
+          val andType = trace(i"andType of scrut", gadts) {
```
Comment (abgruszecki) [inline compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:132]: Ditto about the trace.

File: compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala [hunk 4]
```diff
@@ -94,38 +94,54 @@ trait PatternTypeConstrainer { self: TypeComparer =>
...
+            buildAndType(baseClasses)
+          }
+          constrainPatternType(pat, andType)
```
Comment (abgruszecki) [inline compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:135]: Ditto about `andType` existing.

**Review by abgruszecki — COMMENTED** [review]
Good news, we can merge this PR in time for 3.0.0. Some tweaks would be nice, and then it's good to go!

**Comment by Linyxus** [comment]
Thank you for your review! I have removed unnecessary traces and added the existence checking. It's really good that we can include this feature in 3.0.0. :)

**Review by abgruszecki — APPROVED** [review]

**Pull request merged**